cql3/selection/selection.cc

```diff
@@ -386,17 +388,61 @@ std::unique_ptr<result_set> result_set_builder::build() {
     return std::move(_result_set);
 }
 
-result_set_builder::restrictions_filter::restrictions_filter(::shared_ptr<restrictions::statement_restrictions> restrictions,
+namespace {
+
+using namespace expr;
+
+/// A visitor that copies an expression, but replaces token operator and unselected-column operators with the
+/// TRUE literal.
+struct drop_token_and_unselected_columns {
+    const selection& sel;
+
+    expression operator()(const conjunction& c) const {
+        auto r = c.children | std::views::transform([this] (const expression& e) { return std::visit(*this, e); });
+        return conjunction{std::vector<expression>(r.begin(), r.end())};
+    }
+
+    expression operator()(const binary_operator& binop) const {
+        return seastar::visit(binop.lhs,
+                [] (const token&) { return expression(true); },
+                [&] (const column_value& cv) {
+                    return sel.has_column(*cv.col) ? expression(binop) : expression(true);
+                },
+                [&] (auto&& default_case) { return expression(binop); });
+    }
+
+    expression operator()(bool b) const { return b; }
+};
+
+/// Compiles an expression equivalent to the given restrictions.  A filter can evaluate this expression on each
+/// row to satisfy the restrictions.
+expression compile(
+        const restrictions::statement_restrictions& restrictions,
+        const query_options& options,
+        const selection& sel) {
+    auto ex = restrictions.get_partition_key_restrictions()->expression;
+    if (restrictions.ck_restrictions_need_filtering()) {
+        ex = make_conjunction(std::move(ex), restrictions.get_clustering_columns_restrictions()->expression);
+    }
+    for (auto r : restrictions.get_non_pk_restriction()) {
+        ex = make_conjunction(std::move(ex), r.second->expression);
+    }
+    // TODO: Bind markers here instead of in do_filter, which redoes it for each row.
+    return std::visit(drop_token_and_unselected_columns{sel}, ex);
```

Comment:
I'm puzzled by this due to missing high-level infromation. Here you merge several expressions parts into one. I'd expect the direction to be opposite: the grammar generates one big expression (doesn't today) and the planner works to split it according to what we can execute.

Reply:
I share that vision, but it'll take a few more steps to get there.  Right now I have to deal with multiple expressions coming from upstream, and I want to streamline and optimize their evaluation.  Tomorrow when the upstream hands me a single expression, I'll drop the merging but keep the optimization.

Reply:
Ok.